atomicfu/src/concurrentMain/kotlin/kotlinx/atomicfu/parking/ThreadParker.kt

```diff
@@ -0,0 +1,158 @@
+                            // state hasn't changed so parker is not awake yet, and responsible for cleanup.
+                            myUnparkingState -> if (state.compareAndSet(myUnparkingState, Free)) return
+                            
+                            // Unparked in the meantime (or parker already set state to free and than unpark)
+                            is UnparkedWhileUnparking -> {
+                                
+                                // Parker still asleep. It's parkers responsibility to cleanup.
+                                if (changedState.attempt == myUnparkingState && state.compareAndSet(changedState, Unparked)) return
+                                
+                                // This means that parker was already gone and parker and other thread is unparking.
+                                // Therefore, we are still responsible for cleaning up parking data of our attempt.
+                                if (changedState.attempt != myUnparkingState) {
+                                    delegator.destroyRef(currentState.data)
+                                    return
+                                }
+                            }
+                            
+                            // Parker is already gone. Cleanup
+                            Free -> {
+                                delegator.destroyRef(currentState.data)
+                                return
+                            }
+                            
+                            // Parker is already gone and concurrent unpark call. Cleanup.
+                            Unparked -> {
+                                delegator.destroyRef(currentState.data)
+                                return
+                            }
+                            
+                            // Parker is already gone and parked again in the meantime. Cleanup data of first park.
+                            is Parked -> {
+                                delegator.destroyRef(currentState.data)
+                                return
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+private interface ParkingState
+// The Parker is pre-unparked. The next park call will change state to Free and return immediately.
+private object Unparked : ParkingState
+
+// Starting state. Can be pre unparked or parked from here. A park call will result in Parked state and suspended thread.
+private object Free : ParkingState
+
+// Parker is suspended and can be signaled with data.
+private class Parked(val data: ParkingData) : ParkingState
+
+// An unpark call happened while the parker was parked. In process of unparking.
+private class Unparking : ParkingState
+
+// An unpark call happened while unparking was already in process.
+private class UnparkedWhileUnparking(val attempt: Unparking) : ParkingState
```

Comment:
What problem does this new state solve? Could you provide a non-linearizable interleaving that this state is protecting against?